google-cloud-datastore/src/test/java/com/google/cloud/datastore/it/ITDatastoreTest.java

```diff
@@ -0,0 +1,670 @@
+    assertEquals(ENTITY2, results.next());
+    assertFalse(results.hasNext());
+    transaction.delete(ENTITY3.key());
+    // update entity2 during the transaction
+    DATASTORE.put(Entity.builder(ENTITY2).clear().build());
+    try {
+      transaction.commit();
+      fail("Expecting a failure");
+    } catch (DatastoreException expected) {
+      assertEquals("ABORTED", expected.reason());
+    }
+  }
+
+  @Test
+  public void testNewTransactionRollback() {
+    Transaction transaction = DATASTORE.newTransaction();
+    transaction.add(ENTITY3);
+    Entity entity2 = Entity.builder(ENTITY2).clear().setNull("bla")
+        .set("list3", StringValue.of("bla"), StringValue.builder("bla").build()).build();
+    transaction.update(entity2);
+    transaction.delete(KEY1);
+    transaction.rollback();
+    transaction.rollback(); // should be safe to repeat rollback calls
+
+    try {
+      transaction.commit();
+      fail("Expecting a failure");
+    } catch (DatastoreException ex) {
+      // expected to fail
+    }
+
+    List<Entity> list = DATASTORE.fetch(KEY1, KEY2, KEY3);
+    assertEquals(ENTITY1, list.get(0));
+    assertEquals(ENTITY2, list.get(1));
+    assertNull(list.get(2));
+    assertEquals(3, list.size());
+  }
+
+  @Test
+  public void testNewBatch() {
+    Batch batch = DATASTORE.newBatch();
+    Entity entity1 = Entity.builder(ENTITY1).clear().build();
+    Entity entity2 = Entity.builder(ENTITY2).clear().setNull("bla").build();
+    Entity entity4 = Entity.builder(KEY4).set("value", StringValue.of("value")).build();
+    Entity entity5 = Entity.builder(KEY5).set("value", "value").build();
+
+    List<Entity> entities = batch.add(entity4, PARTIAL_ENTITY2, entity5);
+    Entity entity6 = entities.get(1);
+    assertSame(entity4, entities.get(0));
+    assertEquals(PARTIAL_ENTITY2.names(), entity6.names());
+    assertEquals(PARTIAL_ENTITY2.key().projectId(), entity6.key().projectId());
+    assertEquals(PARTIAL_ENTITY2.key().namespace(), entity6.key().namespace());
+    assertEquals(PARTIAL_ENTITY2.key().ancestors(), entity6.key().ancestors());
+    assertEquals(PARTIAL_ENTITY2.key().kind(), entity6.key().kind());
+    assertEquals(PARTIAL_ENTITY2.key(), IncompleteKey.builder(entity6.key()).build());
+    assertEquals(PARTIAL_ENTITY2.key().ancestors(), entity6.key().ancestors());
+    assertNotEquals(PARTIAL_ENTITY2.key(), entity6.key());
+    assertSame(entity5, entities.get(2));
+    batch.addWithDeferredIdAllocation(PARTIAL_ENTITY3);
+    batch.put(ENTITY3, entity1, entity2);
+
+    Batch.Response response = batch.submit();
+    entities = DATASTORE.fetch(KEY1, KEY2, KEY3, entity4.key(), entity5.key(), entity6.key());
+    assertEquals(entity1, entities.get(0));
+    assertEquals(entity2, entities.get(1));
+    assertEquals(ENTITY3, entities.get(2));
+    assertEquals(entity4, entities.get(3));
+    assertEquals(entity5, entities.get(4));
+    assertEquals(entity6, entities.get(5));
+    assertEquals(6, entities.size());
```

Comment:
This is there just to ensure that there are no more entities than expected. In case an entity is missing we'll fail earlier with an exception.